### PR TITLE
fixed Float16 from Float64 and BigFloat

### PR DESCRIPTION
**Pull request by oscardssmith**
previously `Float16(::Float64)` and `Float16(::BigFloat)` had double rounding issues.

### CODE REVIEW

**Comment by oscardssmith** [comment]
Tests added. This still has a bug for some values less than `nextfloat(Float16(0))`, but it is a major improvement.

File: base/mpfr.jl [hunk 1]
```diff
@@ -338,8 +338,18 @@ Float32(x::BigFloat, r::MPFRRoundingMode=ROUNDING_MODE[]) =
     _cpynansgn(ccall((:mpfr_get_flt,:libmpfr), Float32, (Ref{BigFloat}, MPFRRoundingMode), x, r), x)
 Float32(x::BigFloat, r::RoundingMode) = Float32(x, convert(MPFRRoundingMode, r))
 
-# TODO: avoid double rounding
-Float16(x::BigFloat) = Float16(Float64(x))
+function Float16(x::BigFloat) :: Float16
+    res = Float32(x)
+    resi = reinterpret(UInt32, res)
+    if (resi&0x7fffffff) < 0x38800000
```
Comment (KristofferC) [inline base/mpfr.jl:344]: Maybe this starts to need some comments now. It's a bit of magic number soup now.

Comment (oscardssmith) [reply base/mpfr.jl:344]: good point. that's just if `Float16(res)` is subnormal.

Comment (KristofferC) [reply base/mpfr.jl]: And this is more efficient than `issubnormal(Float16(res))`?

Comment (oscardssmith) [reply base/mpfr.jl]: yeah. this is only 2 instructions while converting to float16 is relatively expensive.

**Comment by oscardssmith** [comment]
Should we merge this as is? This is already a major improvement over current status and I'm not 100% sure how to fix the last remaining case the edge between `floatmin(Float16)` and `0`

File: base/mpfr.jl [hunk 2]
```diff
@@ -338,8 +338,20 @@ Float32(x::BigFloat, r::MPFRRoundingMode=ROUNDING_MODE[]) =
...
-Float16(x::BigFloat) = Float16(Float64(x))
+function Float16(x::BigFloat) :: Float16
+    res = Float32(x)
+    resi = reinterpret(UInt32, res)
+    if (resi&0x7fffffff) < 0x38800000 # if Float16(res) is subnormal
+        #shift so that the mantissa lines up where it would for normal Float16
+        shift = 113-((resi & 0x7f800000)>>23)
+        shift<23 && (resi >>= shift)
+    end
+    if (resi & 0x1fff == 0x1000) # if we are halfway between 2 Float16 values
```
Comment (vtjnash) [inline base/mpfr.jl:349]: The BigFloat `x` (may) have more trailing zeros than this number

**Comment by vtjnash** [comment]
Do you mean all of `0.0:floatmin(Float16)` or just `0.0:nextfloat(Float16(0))`

**Comment by oscardssmith** [comment]
yeah. Fixed.

File: src/runtime_intrinsics.c [hunk 3]
```diff
@@ -11,6 +11,7 @@
 #include "julia.h"
 #include "julia_internal.h"
 #include "APInt-C.h"
+#include "stdio.h"
```
Comment (vchuravy) [inline src/runtime_intrinsics.c:14]: necessary change?

Comment (oscardssmith) [reply src/runtime_intrinsics.c]: Oops, that was there from debugging. Fixed.

**Comment by oscardssmith** [comment]
last remaining bug fixed! from my perspective, this is ready to merge.

**Comment by KristofferC** [comment]
How is it ensured that these tests both test the C version and the Julia version?

**Comment by oscardssmith** [comment]
the c version is float64. the Julia version is bigfloat

**Review by vtjnash — APPROVED** [review]
I helped co-author this, so I will approve here too.

**Comment by oscardssmith** [comment]
Any objections to merging this?

**Comment by KristofferC** [comment]
Needs a manual backport towards 1.7.

**Comment by oscardssmith** [comment]
What branch should I target the PR to?

**Comment by KristofferC** [comment]
`backports-release-1.7`.

**Comment by oscardssmith** [comment]
backported: https://github.com/JuliaLang/julia/pull/43092 is the backport

**Comment by KristofferC** [comment]
Also needs a manual backport towards `backports-release-1.6`.

**Pull request merged**